### DISCUSSION TITLE
Unable to submit content for Review & Publish

### DISCUSSION THREAD

**Discussion by vayu068** [Sunbird-Ed/Community#236] [Installation]
@keshavprasadms, As per the recommendation, creating separate thread for issue being faced in Review & Publish API calls.

1. Using Postman, submitted the uploaded content for Review via content/v1/review/ API & read content from via content/v1/read API. (Got successful results with status set to Review. Still on UI no content is available for under Review section)

![image](https://user-images.githubusercontent.com/8116886/199708470-bc7a839c-f2dc-4852-bd1d-ab0a67365315.png)

2. Tried publishing the content via content/v1/publish API, getting Token Invalid error although using same token, earlier APIs worked.

![image](https://user-images.githubusercontent.com/8116886/199708532-6baa15dc-5905-49fd-9bed-16c4b128c8ff.png)

Please let me know what logs are required to further diagnose the issue.

**Comment by beepdot** [comment]
@Jayaprakash8887 @AmiableAnil Can you folks please answer here?

**Reply by Jayaprakash8887** [reply]
TOKEN validation is not performed by API/Backend service. Please generate a new token with all necessary API permissions and try.

**Reply by Jayaprakash8887** [reply]
For 'Still on UI no content is available for under Review section' issue, chances are Elastic search sync might have not happened. Please sync the object to ES if the content is not available with latest data in content/composite search API

**Reply by vayu068** [reply]
> ntent is available for under Review section' issue, chances are Elastic search sync might have not happened. Please sync the object to ES if the content is not available with latest data in content/composite search API

Please let me know how to do this.


> n with all necessary API permissions and try.

I generated a new access token. With that token, Review API worked but Publish wasn't working. How to check if that token has API permission for Publish. I am generating token with https://<domain_name>/auth/realms/sunbird/protocol/openid-connect/token and as part of body i am sending username, password, client_id as 'portal', client_secret and grant_type being set to password. 
domain_name is set to my working FQDN

**Reply by HarishGangula** [reply]
@vayu068  There are scenarios where this error might occurs
1. When the user token used while calling the publish API is expired
2. If the content creator or collaborator is trying to publish means passing their user token in the publishing call also causes this error

cc: @keshavprasadms @Jayaprakash8887 @AmiableAnil

**Reply by vayu068** [reply]
> @vayu068 There are scenarios where this error might occurs
> 
> 1. When the user token used while calling the publish API is expired
> 2. If the content creator or collaborator is trying to publish means passing their user token in the publishing call also causes this error
> 
> cc: @keshavprasadms @Jayaprakash8887 @AmiableAnil

@HarishGangula I ensured that token isn't expired. I reverified by using the same token to upload & review another content. It worked
2. I have granted all roles to the same user. So access permission shall not be the issue.
![Screenshot 2022-11-16 at 9 53 38 AM](https://user-images.githubusercontent.com/8116886/202083600-5aca4fdf-0245-46b8-9af9-6e9d0b85db91.png)
Please let me know if you require any service logs to diagnose the issue.

**Reply by beepdot** [reply]
Just to be specific. I hope you are not trying to create and publish a content using the same user. Creator and Reviewer should be different user accounts.

**Reply by vayu068** [reply]
Ohh, I was doing it from same account as it was having all roles. I will try publishing from another account.
So, content creation & Review is possible from same account but for publishing another account is required?

**Reply by vayu068** [reply]
@keshavprasadms Thanks for your input, i tried calling publish API from other account. It was successful
<img width="961" alt="Screenshot 2022-11-16 at 4 13 18 PM" src="https://user-images.githubusercontent.com/8116886/202159727-960b2bf4-6872-4484-b3bf-accd5879d8aa.png">
However, i can't see any content in UI. 
I executed the Read API, status was coming as Review only. 
<img width="1014" alt="Screenshot 2022-11-16 at 4 17 39 PM" src="https://user-images.githubusercontent.com/8116886/202160632-83f6db37-d446-415c-86c0-69cc82e763fe.png">
I think publish job is failing. Please let me know what may be wrong here.

**Reply by beepdot** [reply]
In sunbird, a content creator cannot review his own content. It needs to be reviewed and published by another user.
@Jayaprakash8887 @AmiableAnil  Can you please suggest here what to look for in case of publishing failure?

**Reply by Jayaprakash8887** [reply]
@vayu068 Can you pls verify/share the logs in 'content-publish' flink job, if it is running fine. (Check health of job manager and Task manager)

**Reply by vayu068** [reply]
Sorry for the delay, but the log says its unable to connect to content-publish-jobmanager.

`2022-11-24 17:08:55,787 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor           [] - Could not resolve ResourceManager address akka.tcp://flink@content-publish-jobmanager:6123/user/rpc/resourcemanager_*, retrying in 10000 ms: Could not connect to rpc endpoint under address akka.tcp://flink@content-publish-jobmanager:6123/user/rpc/resourcemanager_*.`


Please let me know how to resolve. I don't see any content-publish-jobmanager pod in flink-kp-dev namespace as well.

**Reply by vayu068** [reply]
@keshavprasadms , I would also like to know since my content is in Review status. It should be atleast  be visible in the workspace of creator account. This is also not happening

**Reply by Jayaprakash8887** [reply]
@vayu068 Under what menu are you looking for the content? Is the content read API giving you result. Is the content/composite search API giving you the result when search using 'identifier' ?

**Comment by AmiableAnil** [comment]
@vayu068 Which release branch/tag (learning service and devops repo) are you have build and deployed?

**Reply by vayu068** [reply]
List of tags used for Build (Core Jobs) are
1. Learner service: release-4.6.0_RC6
2. LMS: release-4.6.0_RC3

List of tags used for Deploy (Kubernetes Jobs) are
1. Learner: release-4.6.0_RC4
2. LMS: release-4.6.0_RC4

**Reply by beepdot** [reply]
@vayu068 Are you using `/api` in the URL? Example - `https://domain/api/content/v1/publish/:id`

**Reply by vayu068** [reply]
Yes , {{baseurl}} variable include api

**Reply by beepdot** [reply]
ok, @AmiableAnil does this error `ERR_TOKEN_INVALID` come from any of your services?

**Reply by Jayaprakash8887** [reply]
@keshavprasadms No. There is no token validation in service. Not sure about knowledge middleware. Will ask Harish Gangula to respond regarding token validation in middleware.

**Reply by Krishnaj20** [reply]
@keshavprasadms @sharathkashyap - Is this thread still open?

**Reply by beepdot** [reply]
Yes open. Waiting for inputs from @vayu068
@HarishGangula has provided some info in the previous comment thread.

**Comment by beepdot** [comment]
@vayu068 Is this thread still active? Do you still need any support on this?

**Reply by chitranshu-keshav** [reply]
@vayu068 can you please mark this thread as answered OR get back with the relevant details
